File: android/src/main/java/com/theoplayer/integration/IntegrationRegistration.kt
package com.theoplayer.integration

import androidx.annotation.MainThread
import com.theoplayer.android.api.event.EventType
import com.theoplayer.android.api.event.ads.AdEvent
import com.theoplayer.android.api.event.player.PlayerEvent
import java.io.Closeable

/**
 * Controls event delivery and the lifetime of an Android facade integration.
 *
 * Only the active registration can dispatch events or install interceptors. All operations run on the main thread.
 * Closing a registration restores native delegation without destroying either playback engine.
 */
@MainThread
class IntegrationRegistration internal constructor(
  private val facade: PlayerFacade,
  integration: Integration,
) : Closeable {
  internal var integration: Integration? = integration
    private set

  /**
   * Delivers [event] only to facade player listeners, bypassing native-event interceptors.
   * Does not emit on the raw content player. Calls after this registration closes have no effect.
   */
  fun dispatchEvent(event: PlayerEvent<*>) = facade.dispatchPlayerEvent(this, event)

  /**
   * Delivers [event] only to facade Ads listeners without deriving advertising state from it.
   * Bypasses native-ad filtering. Calls after this registration closes have no effect.
   */
  fun dispatchEvent(event: AdEvent<*>) = facade.dispatchAdEvent(this, event)

  /** Compatibility entry point for player events, equivalent to [dispatchEvent]. */
  fun dispatchPlayerEvent(event: PlayerEvent<*>) = dispatchEvent(event)

  /**
   * Intercepts native events of [type], even without facade listeners, before catch-all interceptors.
   * [interceptor] runs once per event; return true to consume it or false to continue delivery.
   *
   * The returned [Closeable] removes the interceptor. It must be closed explicitly: interception survives
   * unregistering so queued native events can drain. Closed registrations return a no-op disposer.
   */
  fun <T : PlayerEvent<*>> interceptPlayerEvent(type: EventType<T>, interceptor: (T) -> Boolean): Closeable =
    if (facade.isCurrent(this)) facade.playerEvents.intercept(type, interceptor) else Closeable {}

  /**
   * Intercepts all subscribed native player events after type-specific interceptors, including future subscriptions.
   * Return true from [interceptor] to consume the event. Does not subscribe to otherwise unobserved event types.
   *
   * The returned [Closeable] has the same explicit lifetime as [interceptPlayerEvent].
   * Integration-dispatched events bypass interception. Consuming native destruction never prevents facade cleanup.
   */
  fun interceptPlayerEvents(interceptor: (PlayerEvent<*>) -> Boolean): Closeable =
    if (facade.isCurrent(this)) facade.playerEvents.interceptAll(interceptor) else Closeable {}

  /**
   * Restores native playback and Ads delegation without destroying the content player. Idempotent.
   * Existing player-event interceptors remain until their disposers or the facade are closed.
   */
  override fun close() {
    integration = null
    facade.unregister(this)
  }
}

File: src/internal/adapter/web/IntegrationRegistration.ts
import type { PlayerEventMap } from 'theoplayer';
import type { PlayerFacadeEvent } from './PlayerFacadeTypes';

/**
 * Controls event delivery and the lifetime of a registered Web integration.
 *
 * @category Player
 * @platform web
 * @internal
 */
export interface IntegrationRegistration {
  /**
   * Compatibility entry point for player events; equivalent to dispatchEvent(event).
   *
   * @param event - The player event to deliver to facade listeners.
   */
  dispatchPlayerEvent<TEvent extends PlayerEventMap[keyof PlayerEventMap]>(event: TEvent): void;

  /**
   * Routes typed THEOplayer events to player or player.ads listeners without emitting on the content player.
   * Bypasses backing-event interceptors and ignores calls after the registration closes.
   *
   * @param event - The player or advertising event to deliver. Advertising events go only to Ads listeners.
   */
  dispatchEvent<TEvent extends PlayerFacadeEvent>(event: TEvent): void;

  /**
   * Return true to consume a backing-player event before facade listeners receive it.
   * Dispose explicitly: interceptors survive unregistering to allow already-queued events to drain.
   *
   * @param type - The event type to intercept, even when it has no facade listeners.
   * @param interceptor - Called once per event; return `false` to continue delivery.
   * @returns An idempotent disposer. Closed registrations return a no-op disposer.
   */
  interceptPlayerEvent<TType extends keyof PlayerEventMap>(type: TType, interceptor: (event: PlayerEventMap[TType]) => boolean): () => void;

  /**
   * Filters every subscribed backing-player event after type-specific interceptors, including future subscriptions.
   * Uses the same explicit disposer lifetime; integration dispatch bypasses this filter.
   *
   * @param interceptor - Return `true` to consume the event, or `false` to continue delivery.
   * @returns An idempotent disposer. Closed registrations return a no-op disposer.
   *
   * @remarks
   * This does not subscribe to otherwise unobserved event types. Consuming a native destroy event does not prevent facade cleanup.
   */
  interceptPlayerEvents(interceptor: (event: PlayerEventMap[keyof PlayerEventMap]) => boolean): () => void;

  /**
   * Restores content-player delegation without destroying the content player.
   *
   * @remarks
   * Idempotent. Stops integration dispatch and restores native Ads subscriptions.
   * Existing player-event interceptors remain until explicitly disposed or the facade closes.
   */
  close(): void;
}
